Regression in 1.1: `script_path` in `remote-exec` no longer uploads file to `~` paths

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to Terraform Cloud/Enterprise, please contact tf-cloud@hashicorp.support.
If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://registry.terraform.io/browse/providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
❯ terraform -version
Terraform v1.1.2
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.70.0
+ provider registry.terraform.io/hashicorp/http v2.1.0
+ provider registry.terraform.io/hashicorp/tls v3.1.0
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
data "http" "ip" {
  url = "http://ipv4.icanhazip.com"
}

data "aws_ami" "focal" {
  most_recent = true

  filter {
    name   = "virtualization-type"
    values = ["hvm"]
  }

  filter {
    name   = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server*"]
  }

  owners = ["099720109477"]
}

resource "aws_key_pair" "test" {
  key_name_prefix = "test-terraform-provisioner-"
  public_key      = tls_private_key.test.public_key_openssh
}

resource "tls_private_key" "test" {
  algorithm = "RSA"
}

resource "aws_security_group" "test" {
  name_prefix = "test-terraform-provisioner-"

  ingress {
    from_port   = "22"
    to_port     = "22"
    protocol    = "tcp"
    cidr_blocks = ["${chomp(data.http.ip.body)}/32"]
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}

resource "aws_instance" "test" {
  ami                    = data.aws_ami.focal.id
  key_name               = aws_key_pair.test.id
  vpc_security_group_ids = [aws_security_group.test.id]
  instance_type          = "t3.micro"

  connection {
    type        = "ssh"
    host        = self.public_ip
    user        = "ubuntu"
    port        = "22"
    private_key = tls_private_key.test.private_key_pem
  }

  provisioner "remote-exec" {
    inline = [
      "echo foo"
    ]

    connection {
      script_path = "~/foo.sh"
    }
  }
}
```

### Expected Behavior
<!--
What should have happened?
-->

Expected terraform to create the instance, connect to it, and run the inline script. This worked fine in Terraform 1.0, but it fails in Terraform 1.1.

### Actual Behavior
<!--
What actually happened?
-->

Provisioner fails to create the inline script in the specified location. `chmod` executes and resolves the `~`, but then cannot find the file, and the provisioner exits with that error.

```
aws_instance.test (remote-exec): Connected!
╷
│ Error: remote-exec provisioner error
│
│   with aws_instance.test,
│   on main.tf line 62, in resource "aws_instance" "test":
│   62:   provisioner "remote-exec" {
│
│ Failed to upload script: Error chmodding script file to 0777 in remote machine error executing "chmod 0777 ~/foo.sh": Process exited with status 1: chmod: cannot access '/home/ubuntu/foo.sh': No such file or directory
```

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

1. `terraform init`
2. `terraform apply`

Relevant debug output:

```
2021-12-22T08:08:36.576-0800 [DEBUG] Connecting to <public-ip>:22 for SSH
2021-12-22T08:08:36.680-0800 [DEBUG] Connection established. Handshaking for user ubuntu
aws_instance.test (remote-exec): Connected!
2021-12-22T08:08:37.547-0800 [DEBUG] starting ssh KeepAlives
2021-12-22T08:08:37.547-0800 [DEBUG] opening new ssh session
2021-12-22T08:08:38.878-0800 [DEBUG] Starting remote scp process:  'scp' -vt \~
2021-12-22T08:08:38.982-0800 [DEBUG] Started SCP session, beginning transfers...
2021-12-22T08:08:38.982-0800 [DEBUG] Beginning file upload...
2021-12-22T08:08:39.080-0800 [DEBUG] SCP session complete, closing stdin pipe.
2021-12-22T08:08:39.080-0800 [DEBUG] Waiting for SSH session to complete.
2021-12-22T08:08:39.179-0800 [ERROR] scp stderr: "Sink: C0644 19 foo.sh\n"
2021-12-22T08:08:39.179-0800 [DEBUG] opening new ssh session
2021-12-22T08:08:39.391-0800 [DEBUG] starting remote command: chmod 0777 ~/foo.sh
2021-12-22T08:08:39.495-0800 [DEBUG] remote command exited with '1': chmod 0777 ~/foo.sh
2021-12-22T08:08:39.496-0800 [WARN]  Errors while provisioning aws_instance.test with "remote-exec", so aborting
2021-12-22T08:08:39.519-0800 [ERROR] vertex "aws_instance.test" error: remote-exec provisioner error
```

And here's the same section of debug output using terraform 1.0.11:

```
2021-12-22T08:14:28.443-0800 [DEBUG] Connecting to <public-ip>:22 for SSH
2021-12-22T08:14:28.552-0800 [DEBUG] Connection established. Handshaking for user ubuntu
aws_instance.test (remote-exec): Connected!
2021-12-22T08:14:29.378-0800 [DEBUG] starting ssh KeepAlives
2021-12-22T08:14:29.378-0800 [DEBUG] opening new ssh session
2021-12-22T08:14:31.197-0800 [DEBUG] Starting remote scp process:  scp -vt ~
2021-12-22T08:14:31.302-0800 [DEBUG] Started SCP session, beginning transfers...
2021-12-22T08:14:31.302-0800 [DEBUG] Beginning file upload...
2021-12-22T08:14:31.408-0800 [DEBUG] SCP session complete, closing stdin pipe.
2021-12-22T08:14:31.408-0800 [DEBUG] Waiting for SSH session to complete.
2021-12-22T08:14:31.510-0800 [ERROR] scp stderr: "Sink: C0644 19 foo.sh\n"
2021-12-22T08:14:31.510-0800 [DEBUG] opening new ssh session
2021-12-22T08:14:31.730-0800 [DEBUG] starting remote command: chmod 0777 ~/foo.sh
2021-12-22T08:14:31.838-0800 [DEBUG] remote command exited with '0': chmod 0777 ~/foo.sh
2021-12-22T08:14:31.838-0800 [DEBUG] opening new ssh session
2021-12-22T08:14:32.040-0800 [DEBUG] starting remote command: ~/foo.sh
aws_instance.test (remote-exec): foo
2021-12-22T08:14:32.145-0800 [DEBUG] remote command exited with '0': ~/foo.sh
2021-12-22T08:14:32.145-0800 [DEBUG] opening new ssh session
2021-12-22T08:14:32.250-0800 [DEBUG] Starting remote scp process:  scp -vt ~
2021-12-22T08:14:32.351-0800 [DEBUG] Started SCP session, beginning transfers...
2021-12-22T08:14:32.352-0800 [DEBUG] Copying input data into temporary file so we can read the length
2021-12-22T08:14:32.360-0800 [DEBUG] Beginning file upload...
2021-12-22T08:14:32.457-0800 [DEBUG] SCP session complete, closing stdin pipe.
2021-12-22T08:14:32.457-0800 [DEBUG] Waiting for SSH session to complete.
2021-12-22T08:14:32.606-0800 [ERROR] scp stderr: "Sink: C0644 0 foo.sh\n"
aws_instance.test: Creation complete after 25s [id=i-099dc92e451cf4b47]
2021-12-22T08:14:32.620-0800 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/aws/3.70.0/linux_amd64/terraform-provider-aws_v3.70.0_x5 pid=4157
2021-12-22T08:14:32.620-0800 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-12-22T08:14:32.620-0800 [DEBUG] provider: plugin exited
```

For comparison to a success case with terraform 1.1.2, I modified the connection to specify the full path and grabbed the same section of debug output:

```hcl
    connection {
      script_path = "/home/ubuntu/foo.sh"
    }
```

```
2021-12-22T08:23:43.624-0800 [DEBUG] Connecting to <public-ip>:22 for SSH
2021-12-22T08:23:43.728-0800 [DEBUG] Connection established. Handshaking for user ubuntu
aws_instance.test (remote-exec): Connected!
2021-12-22T08:23:44.533-0800 [DEBUG] starting ssh KeepAlives
2021-12-22T08:23:44.533-0800 [DEBUG] opening new ssh session
2021-12-22T08:23:45.920-0800 [DEBUG] Starting remote scp process:  scp -vt /home/ubuntu
2021-12-22T08:23:46.025-0800 [DEBUG] Started SCP session, beginning transfers...
2021-12-22T08:23:46.025-0800 [DEBUG] Beginning file upload...
2021-12-22T08:23:46.124-0800 [DEBUG] SCP session complete, closing stdin pipe.
2021-12-22T08:23:46.125-0800 [DEBUG] Waiting for SSH session to complete.
2021-12-22T08:23:46.235-0800 [ERROR] scp stderr: "Sink: C0644 19 foo.sh\n"
2021-12-22T08:23:46.235-0800 [DEBUG] opening new ssh session
2021-12-22T08:23:46.439-0800 [DEBUG] starting remote command: chmod 0777 /home/ubuntu/foo.sh
2021-12-22T08:23:46.551-0800 [DEBUG] remote command exited with '0': chmod 0777 /home/ubuntu/foo.sh
2021-12-22T08:23:46.551-0800 [DEBUG] opening new ssh session
2021-12-22T08:23:46.754-0800 [DEBUG] starting remote command: /home/ubuntu/foo.sh
aws_instance.test (remote-exec): foo
2021-12-22T08:23:46.856-0800 [DEBUG] remote command exited with '0': /home/ubuntu/foo.sh
2021-12-22T08:23:46.856-0800 [DEBUG] opening new ssh session
2021-12-22T08:23:46.965-0800 [DEBUG] Starting remote scp process:  scp -vt /home/ubuntu
2021-12-22T08:23:47.065-0800 [DEBUG] Started SCP session, beginning transfers...
2021-12-22T08:23:47.066-0800 [DEBUG] Copying input data into temporary file so we can read the length
2021-12-22T08:23:47.075-0800 [DEBUG] Beginning file upload...
2021-12-22T08:23:47.180-0800 [DEBUG] SCP session complete, closing stdin pipe.
2021-12-22T08:23:47.180-0800 [DEBUG] Waiting for SSH session to complete.
2021-12-22T08:23:47.285-0800 [ERROR] scp stderr: "Sink: C0644 0 foo.sh\n"
aws_instance.test: Creation complete after 29s [id=i-07e1923a09c70a8d2]
2021-12-22T08:23:47.304-0800 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-12-22T08:23:47.312-0800 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/aws/3.70.0/linux_amd64/terraform-provider-aws_v3.70.0_x5 pid=4355
2021-12-22T08:23:47.312-0800 [DEBUG] provider: plugin exited
```

--- a/website/docs/language/resources/provisioners/connection.mdx
+++ b/website/docs/language/resources/provisioners/connection.mdx
@@ -102,6 +102,9 @@ block would create a dependency cycle.
   Defaults to 5 minutes.
 
 * `script_path` - The path used to copy scripts meant for remote execution.
+  For more information, see
+  [How Provisioners Execute Remote Scripts](#how-provisioners-execute-remote-scripts)
+  below.
 
 **Additional arguments only supported by the `ssh` connection type:**
 
@@ -123,9 +126,7 @@ block would create a dependency cycle.
 * `host_key` - The public key from the remote host or the signing CA, used to
   verify the connection.
 
-* `target_platform` - The target platform to connect to. Valid values are `windows` and `unix`. Defaults to `unix` if not set.
-
-  If the platform is set to `windows`, the default `script_path` is `c:\windows\temp\terraform_%RAND%.cmd`, assuming [the SSH default shell](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_server_configuration#configuring-the-default-shell-for-openssh-in-windows) is `cmd.exe`. If the SSH default shell is PowerShell, set `script_path` to `"c:/windows/temp/terraform_%RAND%.ps1"`
+* `target_platform` - The target platform to connect to. Valid values are `"windows"` and `"unix"`. Defaults to `"unix"` if not set.
 
 **Additional arguments only supported by the `winrm` connection type:**
 
@@ -136,6 +137,11 @@ block would create a dependency cycle.
 * `use_ntlm` - Set to `true` to use NTLM authentication, rather than default (basic authentication), removing the requirement for basic authentication to be enabled within the target guest. Further reading for remote connection authentication can be found [here](https://docs.microsoft.com/en-us/windows/win32/winrm/authentication-for-remote-connections?redirectedfrom=MSDN).
 
 * `cacert` - The CA certificate to validate against.
+
+Provisioners typically assume that the remote system runs Microsoft Windows
+when using the `winrm` connection type. Behaviors which would vary based on
+the `target_platform` option if using SSH will instead force the
+Windows-specific behavior when using WinRM, unless otherwise specified.
 
 <a id="bastion"></a>
 
@@ -167,3 +173,78 @@ The `ssh` connection also supports the following fields to facilitate connnectio
 * `bastion_certificate` - The contents of a signed CA Certificate. The certificate argument
   must be used in conjunction with a `bastion_private_key`. These can be loaded from
   a file on disk using the [the `file` function](/language/functions/file).
+
+## How Provisioners Execute Remote Scripts
+
+Provisioners which execute commands on a remote system via a protocol such as
+SSH typically achieve that by uploading a script file to the remote system
+and then asking the default shell to execute it. Provisioners use this strategy
+because it then allows you to use all of the typical scripting techniques
+supported by that shell, including preserving environment variable values
+and other context between script statements.
+
+However, this approach does have some consequences which can be relevant in
+some unusual situations, even though this is just an implementation detail
+for typical use.
+
+Most importantly, there must be a suitable location in the remote filesystem
+where the provisioner can create the script file. By default, Terraform
+chooses a path containing a random number using the following patterns
+depending on how `target_platform` is set:
+
+* `"unix"`: `/tmp/terraform_%RAND%.sh`
+* `"windows"`: `C:/windows/temp/terraform_%RAND%.cmd`
+
+In both cases above, the provisioner replaces the sequence `%RAND%` with
+some randomly-chosen decimal digits.
+
+Provisioners cannot react directly to remote environment variables such as
+`TMPDIR` or use functions like `mktemp` because they run on the system where
+Terraform is running, not on the remote system. Therefore if your remote
+system doesn't use the filesystem layout expected by these default paths
+then you can override it using the `script_path` option in your `connection`
+block:
+
+```hcl
+connection {
+  # ...
+  script_path = "H:/terraform-temp/script_%RAND%.sh"
+}
+```
+
+As with the default patterns, provisioners will replace the sequence `%RAND%`
+with randomly-selected decimal digits, to reduce the likelihood of collisions
+between multiple provisioners running concurrently.
+
+If your target system is running Windows, we recommend uses forward slashes
+instead of backslashes, despite the typical convention on Windows, because
+the Terraform language uses backslash as the quoted string escape character.
+
+### Executing Scripts using SSH/SCP
+
+When using the SSH protocol, provisioners upload their script files using
+the Secure Copy Protocol (SCP), which requires that the remote system have
+the `scp` service program installed to act as the server for that protocol.
+
+Provisioners will pass the chosen script path (after `%RAND%`
+expansion) directly to the remote `scp` process, which is responsible for
+interpreting it. With the default configuration of `scp` as distributed with
+OpenSSH, you can place temporary scripts in the home directory of the remote
+user by specifying a relative path:
+
+```hcl
+connection {
+  type = "ssh"
+  # ...
+  script_path = "terraform_provisioner_%RAND%.sh"
+}
+```
+
+-> **Warning:** In Terraform v1.0 and earlier, the built-in provisioners
+incorrectly passed the `script_path` value to `scp` through a remote shell and
+thus allowed it to be subject to arbitrary shell expansion, and thus created an
+unintended opportunity for remote code execution. Terraform v1.1 and later
+will now correctly quote and escape the script path to ensure that the
+remote `scp` process can always interpret it literally. For modules that will
+be used with Terraform v1.0 and earlier, avoid using untrusted external
+values as part of the `script_path` argument.

--- a/website/docs/language/resources/provisioners/file.mdx
+++ b/website/docs/language/resources/provisioners/file.mdx
@@ -52,41 +52,77 @@ resource "aws_instance" "web" {
 
 The following arguments are supported:
 
-* `source` - This is the source file or folder. It can be specified as
-  relative to the current working directory or as an absolute path. This
-  attribute cannot be specified with `content`.
+* `source` - The source file or directory. Specify it either relative to the
+  current working directory or as an absolute path.
+  This argument cannot be combined with `content`.
 
-* `content` - This is the content to copy on the destination. If destination is a file,
-  the content will be written on that file, in case of a directory a file named
-  `tf-file-content` is created. It's recommended to use a file as the destination. A
-  [`template_file`](https://registry.terraform.io/providers/hashicorp/template/latest/docs/data-sources/file) might be referenced in here, or
-  any interpolation syntax. This attribute cannot be specified with `source`.
+* `content` - The direct content to copy on the destination.
+  If destination is a file, the content will be written on that file. In case
+  of a directory, a file named `tf-file-content` is created inside that
+  directory. We recommend using a file as the destination when using `content`.
+  This argument cannot be combined with `source`.
 
-* `destination` - (Required) This is the destination path. It must be specified as an
-  absolute path.
+* `destination` - (Required) The destination path to write to on the remote
+  system. See [Destination Paths](#destination-paths) below for more
+  information.
+
+## Destination Paths
+
+The path you provide in the `destination` argument will be evaluated by the
+remote system, rather than by Terraform itself. Therefore the valid values
+for that argument can vary depending on the operating system and remote access
+software running on the target.
+
+When connecting over SSH, the `file` provisioner passes the given destination
+path verbatim to the `scp` program on the remote host. By default, OpenSSH's
+`scp` implementation runs in the remote user's home directory and so you can
+specify a relative path to upload into that home directory, or an absolute
+path to upload to some other location. The remote `scp` process will run with
+the access level of the user specified in the `connection` block, and so
+permissions may prevent writing directly to locations outside of the home
+directory.
+
+Because WinRM has no corresponding file transfer protocol, for WinRM
+connections the `file` provisioner uses a more complex process:
+
+1. Generate a temporary filename in the directory given in the remote system's
+   `TEMP` environment variable, using a pseudorandom UUID for uniqueness.
+2. Use sequential generated `echo` commands over WinRM to gradually append
+   base64-encoded chunks of the source file to the chosen temporary file.
+3. Use an uploaded PowerShell script to read the temporary file, base64-decode,
+   and write the raw result into the destination file.
+
+In the WinRM case, the destination path is therefore interpreted by PowerShell
+and so you must take care not to use any meta-characters that PowerShell might
+interpret. In particular, avoid including any untrusted external input in
+your `destination` argument when using WinRM, because it can serve as a vector
+for arbitrary PowerShell code execution on the remote system.
+
+Modern Windows systems support running an OpenSSH server, so we strongly
+recommend choosing SSH over WinRM whereever possible, and using WinRM only as
+a last resort when working with obsolete Windows versions.
 
 ## Directory Uploads
 
-The file provisioner is also able to upload a complete directory to the remote machine.
-When uploading a directory, there are a few important things you should know.
+The `file` provisioner can upload a complete directory to the remote machine.
+When uploading a directory, there are some additional considerations.
 
-First, when using the `ssh` connection type the destination directory must already exist.
-If you need to create it, use a remote-exec provisioner just prior to the file provisioner
-in order to create the directory. When using the `winrm` connection type the destination
-directory will be created for you if it doesn't already exist.
+When using the `ssh` connection type the destination directory must already
+exist. If you need to create it, use a remote-exec provisioner just prior to
+the file provisioner in order to create the directory
 
-Next, the existence of a trailing slash on the source path will determine whether the
-directory name will be embedded within the destination, or whether the destination will
-be created. An example explains this best:
+When using the `winrm` connection type the destination directory will be
+created for you if it doesn't already exist.
 
-If the source is `/foo` (no trailing slash), and the destination is `/tmp`, then the contents
-of `/foo` on the local machine will be uploaded to `/tmp/foo` on the remote machine. The
-`foo` directory on the remote machine will be created by Terraform.
+The existence of a trailing slash on the source path will determine whether the
+directory name will be embedded within the destination, or whether the
+destination will be created. For example:
 
-If the source, however, is `/foo/` (a trailing slash is present), and the destination is
-`/tmp`, then the contents of `/foo` will be uploaded directly into `/tmp`.
+* If the source is `/foo` (no trailing slash), and the destination is `/tmp`,
+  then the contents of `/foo` on the local machine will be uploaded to
+  `/tmp/foo` on the remote machine. The `foo` directory on the remote machine
+  will be created by Terraform.
 
-This behavior was adopted from the standard behavior of
-[rsync](https://linux.die.net/man/1/rsync).
-
--> **Note:** Under the covers, rsync may or may not be used.
+* If the source, however, is `/foo/` (a trailing slash is present), and the
+  destination is `/tmp`, then the contents of `/foo` will be uploaded directly
+  into `/tmp`.

--- a/website/docs/language/upgrade-guides/1-1.mdx
+++ b/website/docs/language/upgrade-guides/1-1.mdx
@@ -23,6 +23,7 @@ small number of users, described in the following sections.
 
 * [Terraform requires macOS 10.13 High Sierra or later](#terraform-requires-macos-1013-high-sierra-or-later)
 * [Preparation for removing Azure AD Graph support in the AzureRM Backend](#preparation-for-removing-azure-ad-graph-support-in-the-azurerm-backend)
+* [Interpretation of remote file paths in the `remote-exec` and `file` provisioners](#interpretation-of-remote-file-paths-in-the-remote-exec-and-file-provisioners)
 * [Changes to `terraform graph`](#changes-to-terraform-graph)
 * [Changes to `terraform state mv`](#changes-to-terraform-state-mv)
 * [Provider checksum verification in `terraform apply`](#provider-checksum-verification-in-terraform-apply)
@@ -53,6 +54,66 @@ not set, and so we strongly recommend planning to migrate to Microsoft Graph
 in the near future to prepare for the final removal of Azure AD Graph support
 in a later Terraform release. However, no immediate change is required before
 upgrading to Terraform v1.1.
+
+## Interpretation of remote file paths in the `remote-exec` and `file` provisioners
+
+When using Terraform's built-in `remote-exec` and `file` provisioners, there
+are two situations where Terraform internally uses
+[Secure Copy Protocol](https://en.wikipedia.org/wiki/Secure_copy_protocol)
+(SCP) to upload files to the remote system at a configuration-specified
+location:
+
+* For [the `file` provisioner](/language/resources/provisioners/file),
+  the primary functionality is to upload a file using SCP, and the
+  `destination` argument specifies the remote path where the file is to be
+  written.
+* For [the `remote-exec` provisioner](/language/resources/provisioners/remote-exec),
+  internally the provisioner works by uploading the given scripts to files
+  on the remote system and then executing them. By default the provisioner
+  selects a temporary filename automatically, but a module author can
+  potentially override that location using the `script_path` argument in the
+  associated [`connection` block](https://www.terraform.io/language/resources/provisioners/connection).
+
+If you are not using either of the specific arguments mentioned above, no
+configuration changes will be required to upgrade to Terraform v1.1.
+
+These provisioners both passing the specified remote paths to the `scp` service
+program on the remote system. In Terraform v1.0 and earlier, the provisioners
+were passing the paths to `scp` in a way that was inadvertently subject to
+_shell expansion_. That inadvertently allowed for convenient shorthands
+such as `~/example` and `$HOME/example` to write into the target user's
+home directory, but also offered an undesirable opportunity for accidental
+remote code execution, such as `$(arbitrary-program)`.
+
+In Terraform v1.1 both of the above remote path arguments are passed _verbatim_
+to the remote `scp` service, without any prior shell expansion. For that reason,
+shell-defined expansion tokens such as `~` and environment variable references
+will no longer be evaluated.
+
+By default, the OpenSSH server and the program `scp` together already interpret
+relative paths as relative to the target user's home directory, and so
+module authors can specify relative paths without any special metacharacters
+in order to request uploading into that default location:
+
+```hcl
+  provisioner "file" {
+    source      = "local.txt"
+    destination = "remote.txt"
+  }
+```
+
+If you maintain a module that was depending on expansion of `~/`, `$HOME/`,
+`${HOME}`/ or similar, remove that prefix so that your module instead specifies
+just a relative path.
+
+This is an intentional compatibility regression which we accepted after due
+consideration of
+[the pragmatic exceptions to our compatibility promises](/language/v1-compatibility-promises#pragmatic-exceptions).
+Specifically, this behavior offered an unintended and non-obvious avenue for
+arbitrary code execution on the remote system if either of the above arguments
+were populated from outside input, and an alternative approach is available
+which doesn't have that drawback, and this is therefore justified on security
+grounds.
 
 ## Changes to `terraform graph`
 

